src/volume_mgr.rs

```diff
@@ -130,11 +135,13 @@ where
         const PARTITION_INFO_LBA_START_INDEX: usize = 8;
         const PARTITION_INFO_NUM_BLOCKS_INDEX: usize = 12;
 
-        if self.open_volumes.is_full() {
+        let mut data = self.data.borrow_mut();
```

Comment:
`borrow_mut` may panic.
I'd prefer if this used `try_borrow_mut` and returned an appropriate error in case the RefCell is already borrowed.

(Not only here, but I won't copy the same comment to all calls to `borrow_mut`.)

Reply:
Wouldn't a double-locking indicate a fault in the filesystem code as opposed to anything the user can do anything about?

Reply:
I added a commit with the requested change. Let me know if you do indeed prefer it.

Reply:
Perhaps I'm misunderstanding something? With this pull request, isn't it possible to share a `&VolumeManager` between threads (cores, interrupt context, whatever), so calls to these public functions can happen concurrently?
How does it require a fault in the filesystem code?

Reply:
I thought it was !Sync. That was certainly the intention.

Reply:
Of course! That's what I was missing: I looked at the docs for the crate, and noticed that VolumeManager implements `Send` and `Sync`. But that will no longer be true with this change, as the contained `RefCell` will prevent the auto trait `Sync` from being implemented.

Let me have another look at the code, but indeed I guess that makes the error handling unnecessary.

Reply:
Yeah, RefCell is [expressly !Sync](https://doc.rust-lang.org/stable/std/cell/struct.RefCell.html#impl-Sync-for-RefCell%3CT%3E). So you can't pass a reference to a VolumeManager between threads.

Reply:
Yes, I know - I just didn't make the connection that the auto-trait gets inherited, so `data` being `!Sync` will automatically cause `VolumeManager` to also be `!Sync`.

Indeed, considering that, currently I don't see a way for multiple borrows to happen at the same time.
The only function that both borrows `data` and calls another function that also borrows `data` is `close_file`, and that one calls `flush_file` before borrowing `data`, so it should be fine.

And I agree, if a future change of the crate violates that property and somehow does recursive calls to `borrow_mut` / `borrow`, that would be a severe bug that could justify a panic.

I'm a bit indecisive whether a panic is _always_ the right response. The firmware may be prepared to continue running after filesystem errors anyway (the SD card may be a non-essential part). In that case, returning a `LockError` could convert a catastrophic (but obvious) bug into a minor inconvenience. At the same time it would hide the underlying bug, and the end-user may not get any useful indication why the appliance stops accessing the file system.

Overall, I'm fine with both variants. With the `shell` example, I don't see a size difference.
Do you prefer panicking version?

Reply:
> Indeed, considering that, currently I don't see a way for multiple borrows to happen at the same time.

I think that maybe you could try and open a file from the iterate closure.

Reply:
Yes I can confirm that trying to do anything with the VolumeManager from inside the iterate closure produces a lock error. So returning the error is better than panicking.

Reply:
Good catch, I didn't notice that possibility. Perhaps in this case you want to reconsider the doc comment on LockError? "This is a bug in the filesystem. Please open an issue."

Reply:
Added some better notes, and a warning on the relevant closure-based APIs.